### PR TITLE
Implement hash collections

### PR DESCRIPTION
**Pull request by ZiyanShi**
Signed-off-by: ZiyanShi <ziyan.shi@intel.com>
Implemented Hashes.
Rename some variables to increase readability.
Add Stress test for hashes.
Split benchmark scripts for easier use.
Change some function signatures to align with RocksDB.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,18 +9,23 @@ include(${KVDK_ROOT_DIR}/cmake/functions.cmake)
...
 set(CMAKE_CXX_STANDARD 17)
-set(OPT "-msse4 -mavx -mavx2 -O2 -fno-omit-frame-pointer -momit-leaf-frame-pointer")
-
 option(DEBUG_CHECK "check connection" OFF)
 if(DEBUG_CHECK)
     ADD_DEFINITIONS(-DDEBUG_CHECK)
-endif()    
+endif()
+
+set(OPT "-mavx -mavx2 -mavx512f")
```
Comment (sherry-1001) [inline CMakeLists.txt:17]: why remove `-fno-omit-frame-pointer -momit-leaf-frame-pointer`?

File: engine/unordered_collection.cpp [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+EmplaceReturn UnorderedCollection::EmplaceBefore(
+    DLDataEntry *pmp, std::uint64_t timestamp, pmem::obj::string_view const key,
+    pmem::obj::string_view const value, DataEntryType type,
+    std::unique_lock<SpinMutex> const &lock) {
+  _CheckUserSuppliedPmp_(pmp);
+  DListIterator iter_prev{_dlinked_list_._p_pmem_allocator_, pmp};
+  --iter_prev;
+  DListIterator iter_next{_dlinked_list_._p_pmem_allocator_, pmp};
+  EmplaceReturn ret =
+      _EmplaceBetween_(iter_prev._pmp_curr_, iter_next._pmp_curr_, timestamp,
```
Comment (JiayuZzz) [inline engine/unordered_collection.cpp:92]: do not use '_' on function name prefix&suffix

File: engine/unordered_collection.cpp [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+namespace KVDK_NAMESPACE {
+UnorderedCollection::UnorderedCollection(
+    std::shared_ptr<PMEMAllocator> sp_pmem_allocator,
+    std::shared_ptr<HashTable> sp_hash_table, std::string const &name,
+    std::uint64_t id, std::uint64_t timestamp) try : _sp_hash_table_ {
+  sp_hash_table
+}
+, _p_pmem_allocator_{sp_pmem_allocator.get()}, _pmp_dlist_record_{nullptr},
+    _dlinked_list_{sp_pmem_allocator, timestamp, _ID2View_(id),
```
Comment (JiayuZzz) [inline engine/unordered_collection.cpp:11]: only suffix '_' on private variable

File: engine/unordered_collection.cpp [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+      return;
+    }
+    case DataEntryType::DlistHeadRecord:
+    case DataEntryType::DlistRecord:
+    default: {
+      goto FATAL_FAILURE;
+    }
+    }
+  }
+FATAL_FAILURE:
```
Comment (JiayuZzz) [inline engine/unordered_collection.cpp:293]: This is not fatal, do not throw, just make iterator invalid is enough

File: engine/unordered_collection.cpp [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+      return;
+    }
+    case DataEntryType::DlistTailRecord:
+    case DataEntryType::DlistRecord:
+    default: {
+      goto FATAL_FAILURE;
+    }
+    }
+  }
+FATAL_FAILURE:
```
Comment (JiayuZzz) [inline engine/unordered_collection.cpp:338]: same as _Next_()

File: engine/unordered_collection.hpp [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+#include "structures.hpp"
+#include "utils.hpp"
+
+namespace KVDK_NAMESPACE {
+/// [Obsolete?]
+// Triplet of unique_kock. Owns all or none of three locks.
+// The locks may duplicate. Duplicate locks are ignored.
+// e.g. if third and first lock are same, third lock is ignored.
+// Locks with same address are considered duplicate.
+template <typename Lock> class UniqueLockTriplet {
```
Comment (JiayuZzz) [inline engine/unordered_collection.hpp:24]: is this class ever used?

**Review by JiayuZzz — APPROVED** [review]
LGTM, thx

**Pull request merged**